File: Day5.cs
using NPOI.HSSF.Record.Chart;
using NPOI.SS.Formula.PTG;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

// Port-over from day5.py
// C# is approx 20 times faster than Python.
// Python will require several hours for N=1e9, whereas
// C# only requires 20min.

namespace Playground
{
    using longerval = Tuple<long, long>;
    using Mapping = Tuple<long, long, long>;
    public class Day5
    {
        private string[] Data;
        public Day5(string path)
        {
            this.Data = File.ReadAllLines(path);
        }

        public void Solve()
        {
            List<longerval> listOfSeeds = ProcessNewSeeds(Data[0]);
            List<List<Mapping>> listOfRawMaps = ProcessMaps(Data.SubArray(2, Data.Length - 2));
            List<Map> listOfMaps = new List<Map>();

            foreach(var map in listOfRawMaps)
            {
                listOfMaps.Add(new Map(map));
            }

            long minLocation = long.MaxValue;
            foreach(var seedRange in listOfSeeds)
            {
                var start = DateTime.Now;

                for(long i = seedRange.Item1; i <= seedRange.Item2; i++)
                {
                    long curr = i;
                    foreach(var map in listOfMaps)
                    {
                        curr = map.Convert(curr);
                    }
                    minLocation = Math.Min(minLocation, curr);
                }
                var end = DateTime.Now;
                Console.WriteLine($"{seedRange.Item2-seedRange.Item1} entries took {end-start} seconds: {minLocation}");
            }

            Console.WriteLine($"Solution is {minLocation}");
        }

        private List<longerval> ProcessNewSeeds(string row)
        {
            string num_list_str = row.Split(":")[1].Trim();
            string[] rawNums = num_list_str.Split(" ");

            List<long> listOfNums = new List<long>();
            foreach (var rawNum in rawNums)
                listOfNums.Add(long.Parse(rawNum));

            List<longerval> listOflongervals = new List<longerval>();
            for(int i = 0; i < listOfNums.Count; i += 2)
            {
                listOflongervals.Add(new longerval(listOfNums[i], listOfNums[i] + listOfNums[i + 1] - 1));
            }

            return listOflongervals;
        }

        private List<List<Mapping>> ProcessMaps(string[] data)
        {
            List<List<Mapping>> listOfMaps = new List<List<Mapping>>();
            List<Mapping> mappings = new List<Mapping>();

            foreach (var s in data)
            {
                var row = s.Trim();
                if (row.Length == 0)
                {
                    listOfMaps.Add(mappings);
                    continue;
                }

                if (row.EndsWith(":"))
                {
                    mappings = new List<Mapping>();
                }
                else
                {
                    var tmp = row.Split(" ");
                    Mapping mapping = new Mapping(long.Parse(tmp[0].Trim()),
                                                  long.Parse(tmp[1].Trim()),
                                                  long.Parse(tmp[2].Trim()));
                    mappings.Add(mapping);
                }
            }

            if (mappings.Count > 0)
                listOfMaps.Add(mappings);
            return listOfMaps;
        }
    }

    public class Map
    {
        private List<Mapping> Mappings;
        public Map(List<Mapping> rawMaps)
        {
            Mappings = new List<Mapping>();
            foreach(Mapping map in rawMaps)
            {
                var (dest, src, dist) = map;
                var entry = new Mapping(src, src + dist - 1, dest);
                Mappings.Add(entry);
            }

            Mappings.Sort();
        }

        public long Convert(long srcId)
        {
            int idx = BinarySearch(0, Mappings.Count - 1, srcId);
            if (idx == -1)
                return srcId;
            else
            {
                var (src, _, dest) = Mappings[idx];
                return dest + (srcId - src);
            }
        }

        private int BinarySearch(int l, int r, long toFind)
        {
            if (l > r) return -1;

            int m = l + (r - l) / 2;
            long start = Mappings[m].Item1;
            long end = Mappings[m].Item2;

            if (start <= toFind && toFind <= end) return m;

            if (toFind < start) return BinarySearch(l, m - 1, toFind);
            else if (toFind > end) return BinarySearch(m + 1, r, toFind);

            return -1;
        }
    }

    public static class Extensions
    {
        public static T[] SubArray<T>(this T[] array, int offset, int length)
        {
            return array.Skip(offset).Take(length).ToArray();
        }
    }
}

File: day5.py
from util import solve

# Do not run this for Part 2, as it will take too long.
# Alternative algorithm is needed, such as search by ranges, etc.
# C# is used to speed up Part 2 algorithm.

def process_seeds(row):
    seed_list_str = row.split(":")[1].strip()

    list_of_seeds = [int(s) for s in seed_list_str.split(" ")]
    return list_of_seeds

def process_new_seeds(row):
    num_list_str = row.split(":")[1].strip()
    list_of_nums = [int(s) for s in num_list_str.split(" ")]

    list_of_intervals = []
    for i in range(0,len(list_of_nums), 2):
        list_of_intervals.append((list_of_nums[i], list_of_nums[i]+list_of_nums[i+1]-1))
    
    return merge_intervals(list_of_intervals)


def merge_intervals(list_of_intervals):
    new_list = sorted(list_of_intervals)
    tmp = []
    for interval in new_list:
        curr = interval
        if not tmp: 
            tmp.append(curr)
            continue

        while is_overlapping(tmp[-1], curr):
            start, end = tmp.pop()
            new_start = min(start, curr[0])
            new_end = max(end, curr[1])

            curr = (new_start, new_end)
        
        tmp.append(curr)
    return tmp

def is_overlapping(i1, i2):
    return i1[0] <= i2[1] and i1[1] >= i2[0]

def process_maps(data: list[str]) -> list[list[str]]:
    list_of_maps = []
    mappings = []
    
    for row in data:
        row = row.strip()
        if not row:
            list_of_maps.append(mappings)
            continue
        
        if row.endswith(":"):
            mappings = []
        else:
            tmp = row.split(" ")
            mapping = (int(tmp[0].strip()), int(tmp[1].strip()), int(tmp[2].strip()))
            mappings.append(mapping)
    
    if mappings:
        list_of_maps.append(mappings)

    return list_of_maps

class Mapping:
    def __init__(self, raw_maps):
        self.mappings = []

        for (dest, src, dist) in raw_maps:
            entry = (src, src+dist-1, dest)
            self.mappings.append(entry)
        self.mappings.sort()
    
    def convert(self, src_id):
        idx = self._binary_search(0, len(self.mappings)-1, src_id)

        if idx == -1:
            return src_id
        else:
            src, _, dest = self.mappings[idx]
            return dest + (src_id-src)
    
    def _binary_search(self, l, r, to_find):
        if l > r:
            return -1

        m = l + (r-l)//2

        start, end = self.mappings[m][0], self.mappings[m][1]
        if start <= to_find and to_find <= end:
            return m
        
        if to_find < start:
            return self._binary_search(l, m-1, to_find)
        elif to_find > end:
            return self._binary_search(m+1, r, to_find)


def solver(data):
    list_of_seeds = process_seeds(data[0])
    list_of_raw_maps = process_maps(data[2:])
    list_of_maps = [Mapping(raw_map) for raw_map in list_of_raw_maps]

    min_location = float("inf")
    for seed in list_of_seeds:
        curr = seed
        for map in list_of_maps:
            curr = map.convert(curr)
        min_location = min(curr, min_location)
    
    return min_location

from time import time
def solver_new(data):
    list_of_seeds = process_new_seeds(data[0])
    list_of_raw_maps = process_maps(data[2:])
    list_of_maps = [Mapping(raw_map) for raw_map in list_of_raw_maps]

    min_location = float("inf")
    list_of_seeds = [(0, 10_000_000)]
    for (start, end) in list_of_seeds:
        begin = time()
        for seed in range(start, end+1):
            curr = seed
            for map in list_of_maps:
                curr = map.convert(curr)
            min_location = min(curr, min_location)
        fin = time()
        print(f"{end-start} entries took {fin-begin} seconds")
    
    return min_location

solve(solver_new, 5)